File: python/leetcode/kth-smallest-element-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def kthSmallest(self, root: TreeNode, k: int) -> int:
        return self.traverse(root)[k - 1]
    def traverse(self, root):
        if root is None:
            return []
        return self.traverse(root.left) + [root.val] + self.traverse(root.right)

# Can also be written as such with arr passed in

# class Solution:
#     def kthSmallest(self, root: TreeNode, k: int) -> int:
#         return self.traverse(root, [])[k - 1]
#     def traverse(self, root, arr):
#         if root is not None:
#             self.traverse(root.left, arr)
#             arr.append(root.val)
#             self.traverse(root.right, arr)
#         return arr

File: java/leetcode/kth-smallest-element-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int kthSmallest(TreeNode root, int k) {
        List<Integer> values = traverse(root, new ArrayList<Integer>());
        return values.get(k - 1);
    }
    public List<Integer> traverse(TreeNode root, ArrayList<Integer> arr) {
        if (root == null) {
            return arr;
        }
        traverse(root.left, arr);
        arr.add(root.val);
        traverse(root.right, arr);
        return arr;
    }
}
